junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,544 @@
+package org.scalajs.junit.plugin
+
+import scala.language.reflectiveCalls
+
+import scala.reflect.internal.Flags
+import scala.tools.nsc._
+import scala.tools.nsc.plugins.{
+  Plugin => NscPlugin, PluginComponent => NscPluginComponent
+}
+
+/** The Scala.js jUnit plugin is a way to overcome make class annotation
+ *  information (usually accessed by reflection) of any test class accessible
+ *  to the Scala.js testing framework.
+ *
+ *  As an example we take the following test class:
+ *  {{{
+ *  class Foo {
+ *    @Before def before(): Unit = {
+ *      // Initialize the instance before the tests
+ *    }
+ *    @Test def bar(): Unit = {
+ *      // assert some stuff
+ *    }
+ *    @Ignore("baz not implemented yet") @Test def baz(): Unit = {
+ *      // assert some other stuff
+ *    }
+ *  }
+ *
+ *  object Foo {
+ *    @BeforeClass def beforeClass(): Unit = {
+ *      // Initialize some global state for the tests.
+ *    }
+ *  }
+ *  }}}
+ *
+ *  Will generate the following bootstrapper module:
+ *
+ *  {{{
+ *  object Foo\$scalajs\$junit\$bootstrapper extends org.scalajs.junit.JUnitTestBootstrapper {
+ *
+ *    def metadata(): JUnitClassMetadata = {
+ *      new JUnitClassMetadata(
+ *        classAnnotations = List(),
+ *        moduleAnnotations = List(),
+ *        classMethods = List(
+ *            new JUnitMethodMetadata(name = "before",
+ *                annotations = List(new Before)),
+ *            new JUnitMethodMetadata(name = "bar",
+ *                annotations = List(new Test)),
+ *            new JUnitMethodMetadata(name = "baz",
+ *                annotations = List(new Test, new Ignore("baz not implemented yet")))
+ *        ),
+ *        moduleMethods(
+ *            new JUnitMethodMetadata(name = "beforeClass",
+ *                annotations = List(new BeforeClass)))
+ *      )
+ *    }
+ *
+ *    def newInstance(): AnyRef = new Foo()
+ *
+ *    def invoke(methodName: String): Unit = {
+ *      if (methodName == "0") Foo.beforeClass()
+ *      else throw new NoSuchMethodException(methodId)
+ *    }
+ *
+ *    def invoke(instance: AnyRef, methodName: String): Unit = {
+ *      if (methodName == "before") instance.asInstanceOf[Foo].before()
+ *      else if (methodName == "bar") instance.asInstanceOf[Foo].bar()
+ *      else if (methodName == "baz") instance.asInstanceOf[Foo].baz()
+ *      else throw new NoSuchMethodException(methodId)
+ *    }
+ *  }
+ *  }}}
+ *  The test framework will identify `Foo\$scalajs\$junit\$bootstrapper` as a test module
+ *  because it extends `JUnitTestBootstrapper`. It will know which methods to run based
+ *  on the info returned by Foo\$scalajs\$junit\$bootstrapper.metadata,
+ *  it will create new test instances using `Foo\$scalajs\$junit\$bootstrapper.newInstance()`
+ *  and it will invoke test methods using `scalajs\$junit\$invoke` methods on the
```

Comment:
`scalajs\$junit\$invoke` is obsolete, here. It should just be `invoke`, I think.